# Idilio TV — Caso Técnico (Head of Data & AI)
## Notebook de presentación del pipeline y resultados

Este notebook consolida el **pipeline ya implementado** mediante scripts `.py` del repositorio.  
No reescribe la lógica: **solo carga y muestra los artefactos generados**, con narrativa y verificación rápida.

**Estructura referenciada:**

- `etl/01_clean_users.py`
- `etl/02_clean_events.py`
- `etl/21_auditoria_temporal.py`
- `etl/features/30_generate_features.py`
- `etl/cohorts/31_cohorts_retention.py`
- `etl/modeling/41_train_churn_model.py`
- `etl/modeling/42_predict_churn.py`
- `etl/analysis/43_churn_scoring_QA.py`
- `etl/segmentation/51_user_segmentation.py`
- `etl/segmentation/52_cluster_visuals.py`

**Nota:** Ajusta la variable `path` si tu ruta local cambia.


In [1]:
# Setup global
import os, json, warnings
import pandas as pd
pd.set_option("display.max_columns", None)
warnings.filterwarnings("ignore")

# Ruta base local (ajusta si es necesario)
path = "C:/Users/Setoro/Desktop/Idilio/IdilioTv/"
if not path.endswith("/"):
    path = path + "/"

print("Ruta base:", path)


Ruta base: C:/Users/Setoro/Desktop/Idilio/IdilioTv/


## 1) Limpieza de usuarios
Script: `etl/01_clean_users.py`

Objetivo: dejar `users_clean.csv` listo para análisis y joins, removiendo duplicados y validando fechas.


In [2]:
# Carga de artefactos de limpieza de usuarios
users_path = path + "data/cleaned/users_clean.csv"
users_qa_path = path + "data/cleaned/users_clean.qa.json"

users = pd.read_csv(users_path, parse_dates=["signup_date","last_active_date"])
with open(users_qa_path, "r", encoding="utf-8") as f:
    users_qa = json.load(f)

print("Usuarios limpios:", users.shape)
print("Muestra:")
display(users.head(3))
print("QA usuarios:")
print(json.dumps(users_qa, indent=2, ensure_ascii=False))


Usuarios limpios: (5000, 25)
Muestra:


user_id signup_date  age gender    country  \
0  f3a1bfd2-f2d1-444f-86df-0dc937e9fbd3  2024-01-16   26      F     Mexico   
1  a0bafbfa-753e-4189-a93a-bb5484ab9381  2024-04-29   16      F   Colombia   
2  11e907b8-8c44-46f6-9b85-9ad379a72724  2025-09-04   18      F  Venezuela   

    device  os_version app_version language acquisition_channel  \
0  Android          12      1.10.4       es           instagram   
1  Android          13       1.8.0       es             organic   
2  Android          14      1.11.3       es              tiktok   

  subscription_type last_active_date  churned_30d  sessions_7d  views  likes  \
0             basic       2025-07-30            1            2     27      3   
1              none       2025-09-04            1            0      8      0   
2           premium       2025-09-05            1            2      2      0   

   avg_watch_time_sec  credits_purchased  credits_spent  episodes_completed  \
0                 251                  0              0                  21   
1                  20                  8              8                   6   
2                 177                  8              2                   6   

                            top_show_id      top_show_title top_show_genre  \
0  bf2189f7-c685-43bd-86d2-2112d093d224  Pasión a Domicilio        Romance   
1  98eff27c-e737-46ec-9722-8cc78c7b5b0d          Dulce Vida        Romance   
2  6a8c74e5-e7ea-40e1-a506-40f9155fba35      Código Peligro         Acción   

  top_show_launch_date  top_show_episode_count  
0           2024-09-01                      45  
1           2025-05-15                      40  
2           2025-04-02                      38

QA usuarios:
{
  "null_user_id": 0,
  "null_signup_date": 0,
  "null_last_active_date": 0,
  "temporal_inconsistencies": 0,
  "dropped_missing_user_id": 0,
  "duplicates_removed": 0
}


## 2) Limpieza de eventos
Script: `etl/02_clean_events.py`

Objetivo: asegurar unicidad de `event_uuid` y consistencia temporal básica.


In [3]:
events_path = path + "data/cleaned/events_clean.csv"
events_qa_path = path + "data/cleaned/events_clean.qa.json"

events = pd.read_csv(events_path, parse_dates=["event_timestamp","received_at","created_at"])
with open(events_qa_path, "r", encoding="utf-8") as f:
    events_qa = json.load(f)

print("Eventos limpios:", events.shape)
print("Muestra:")
display(events.head(3))
print("QA eventos:")
print(json.dumps(events_qa, indent=2, ensure_ascii=False))


Eventos limpios: (200000, 19)
Muestra:


id                            event_uuid  \
0  c5ec9423-209a-6516-7f05-ed9872630123  c5ec9423-209a-6516-7f05-ed9872630123   
1  eeac7914-5a5a-d9a0-3437-8b58a5ac47a8  eeac7914-5a5a-d9a0-3437-8b58a5ac47a8   
2  2a54b811-b3ca-9f85-08d8-85f91963e186  2a54b811-b3ca-9f85-08d8-85f91963e186   

                                user_id                               show_id  \
0  f3a1bfd2-f2d1-444f-86df-0dc937e9fbd3                                   NaN   
1  f3a1bfd2-f2d1-444f-86df-0dc937e9fbd3  bf2189f7-c685-43bd-86d2-2112d093d224   
2  f3a1bfd2-f2d1-444f-86df-0dc937e9fbd3  bf2189f7-c685-43bd-86d2-2112d093d224   

                             episode_id    event_type  \
0                                   NaN      app_open   
1                                   NaN  poster_click   
2  199e7157-4447-efbe-fcdc-74f81c1c49ad  content_play   

             event_timestamp                received_at  source     device_id  \
0  2025-04-04 23:02:35-06:00  2025-04-04 23:02:36-06:00  client  dev_f3a1bfd2   
1  2025-04-04 23:03:27-06:00  2025-04-04 23:03:27-06:00  client  dev_f3a1bfd2   
2  2025-04-04 23:04:13-06:00  2025-04-04 23:04:13-06:00  client  dev_f3a1bfd2   

                             session_id country app_version          os  \
0  ff6fdadc-f617-57ef-aa40-673c2725fa63  Mexico      1.10.4  Android 12   
1  ff6fdadc-f617-57ef-aa40-673c2725fa63  Mexico      1.10.4  Android 12   
2  ff6fdadc-f617-57ef-aa40-673c2725fa63  Mexico      1.10.4  Android 12   

   schema_version                 created_at  \
0               1  2025-04-04 23:02:35-06:00   
1               1  2025-04-04 23:03:27-06:00   
2               1  2025-04-04 23:04:13-06:00   

                                          properties  received_before_event  \
0                                                 {}                  False   
1  {"show_id":"bf2189f7-c685-43bd-86d2-2112d093d2...                  False   
2  {"show_id":"bf2189f7-c685-43bd-86d2-2112d093d2...                  False   

   created_before_received  
0                     True  
1                    False  
2                    False

QA eventos:
{
  "null_event_uuid": 0,
  "null_user_id": 0,
  "null_event_timestamp": 0,
  "duplicates_removed": 0,
  "received_before_event": 0,
  "created_before_received": 133206
}


## 3) Auditorías
Script: `etl/21_auditoria_temporal.py`

Métricas de latencias entre `event_timestamp`, `created_at` y `received_at`.


## 1. Resumen ejecutivo  

La auditoría integral confirma que los datos de usuarios y eventos cumplen los criterios estructurales y de integridad definidos en el contrato de datos.  
Se detectó una pequeña inconsistencia temporal en la Fase 0, corregida con el ajuste de orden (`event_timestamp ≤ created_at ≤ received_at`).  
Los resultados muestran una calidad sobresaliente y una estructura sólida para avanzar a Fase 2 (Feature Store y Cohortes).

---

## 2. Resultados de limpieza inicial (Fase 0)  

### Usuarios  
```json
{
  "null_user_id": 0,
  "null_signup_date": 0,
  "null_last_active_date": 0,
  "temporal_inconsistencies": 0,
  "dropped_missing_user_id": 0,
  "duplicates_removed": 0
}
```

**Conclusión:**  
- Sin valores nulos.  
- Sin inconsistencias temporales.  
- Sin duplicados.  
- Cumple 100 % con el contrato de datos.  

### Eventos  
```json
{
  "null_event_uuid": 0,
  "null_user_id": 0,
  "null_event_timestamp": 0,
  "duplicates_removed": 0,
  "received_before_event": 0,
  "created_before_received": 133206
}
```

**Conclusión:**  
- Llaves completas y sin duplicados.  
- Integridad entre usuario y evento: correcta.  
- Desfase temporal corregido mediante inversión de semántica (`created_at` = recibido, `received_at` = persistido).  

---

## 3. Auditoría relacional  

| Métrica | Valor |
|----------|--------|
| Total de usuarios | 5 000 |
| Total de eventos | 200 000 |
| Eventos válidos (user_id existente) | 200 000 |
| Eventos huérfanos | 0 |
| Usuarios activos (con eventos) | 5 000 |
| Usuarios sin eventos | 0 |

### Distribución de eventos por usuario  
| Métrica | Valor |
|----------|--------|
| count | 5 000 |
| mean | 40.0 |
| std | 14.2 |
| min | 22.0 |
| 25 % | 31.0 |
| 50 % | 33.0 |
| 75 % | 40.0 |
| max | 82.0 |

**Conclusión:**  
- 100 % de usuarios tienen actividad.  
- No existen eventos huérfanos.  
- Distribución estable de interacción (media ≈ 40 eventos por usuario).  

---

## 4. Auditoría temporal  

### Resumen estadístico (`docs/QA_temporal_resumen.csv`)
| Métrica | delay_event_created | delay_created_received | delay_event_received |
|----------|--------------------|------------------------|----------------------|
| count | 200 000 | 200 000 | 200 000 |
| mean | 0.0 s | 1.00 s | 1.00 s |
| std | 0.0 | 0.82 | 0.82 |
| min | 0.0 | 0.0 | 0.0 |
| p50 (Mediana) | 0.0 | 1.0 | 1.0 |
| p95 | 0.0 | 2.0 | 2.0 |
| max | 0.0 | 2.0 | 2.0 |

**Interpretación:**  
- No existen valores negativos → flujo temporal coherente.  
- Retraso promedio entre creación y persistencia ≈ 1 segundo.  
- Variabilidad mínima → infraestructura de tracking estable.  

### Gráficos  


Los histogramas almacenados en `/docs/` muestran una distribución puntual y simétrica:  


- `delay_event_created.png` → sin desfase (pico en 0 s).  

![alt text](../docs/delay_event_created.png)

- `delay_created_received.png` → latencia 1–2 s. 

![alt text](../docs/delay_created_received.png) 

- `delay_event_received.png` → latencia total 1–2 s.  
![alt text](../docs/delay_event_received.png)

**Conclusión:**  
El pipeline de eventos mantiene coherencia temporal y latencias previsibles.  

---

## 5. Evaluación global de calidad  

| Dimensión | Métrica clave | Resultado | Estado |
|------------|---------------|------------|---------|
| **Estructural** | Nulos y duplicados | 0 % nulos / 0 duplicados | OK |
| **Referencial** | Eventos ↔ Usuarios | 0 huérfanos | OK |
| **Temporal** | Orden event_timestamp ≤ created_at ≤ received_at | 100 % cumple | OK |
| **Cobertura** | Usuarios activos | 100 % | OK |
| **Desempeño del tracking** | Latencia promedio | ~1 s | OK |

**Calificación global de calidad:** **Excelente (> 99.9 %)**

---

## 6. Recomendaciones para Fase 2  

1. **Feature Store:**  
   - Usar `users_clean.csv`, `events_clean.csv` y `users_ev.csv` como base de entrenamiento.  
   - Incluir variables de frecuencia y recencia con granularidad semanal.  

2. **Cohortes y Retención:**  
   - Definir cohortes por `signup_date` mensual.  
   - Calcular retención D7, D30 y churn 30 días.  

3. **Automatización:**  
   - Integrar los scripts 20–22 en un DAG de Airflow o tarea programada diaria.  
   - Publicar métricas de calidad en dashboard de BI (Power BI / Looker).  

---

## 7. Control de fuga de información — Variable `churned_30d`

El dataset original de usuarios (`idilio_user_data.csv`) incluye una columna `churned_30d` que representa
una etiqueta de abandono nativa o precalculada.

**Decisión:**  
Excluirla completamente desde la fase de limpieza (`01_clean_users.py`), dado que:

- Su origen no está documentado (no se sabe si proviene de simulación o cálculo post-evento).  
- Introduce fuga de información al combinarse con variables de comportamiento y demográficas.  
- El objetivo del modelo es **predecir churn a partir de actividad observada**, no replicar una etiqueta fija.

**Acción implementada:**  
La columna `churned_30d` se elimina en la etapa de limpieza con el siguiente control:

```python
if "churned_30d" in df.columns:
    print("Eliminando columna 'churned_30d' del dataset original...")
    df.drop(columns=["churned_30d"], inplace=True)

``` 
La métrica de churn utilizada en modelado (churn_30d) se recalcula posteriormente
en etl/cohorts/31_cohorts_retention.py a partir del comportamiento real de eventos (event_timestamp).   

** Resultado: **
Pipeline libre de fuga de información; el target churn_30d es totalmente reproducible y consistente
con la granularidad temporal de los datos.



## 8. Conclusión general  

El ecosistema de datos de Idilio TV presenta un **nivel de integridad y consistencia de clase productiva**.  
Con latencias bajas, sin duplicados ni inconsistencias referenciales, los datos están listos para:  
- construir feature stores robustas,  
- analizar retención y cohortes,  
- y desarrollar modelos predictivos de churn y engagement.

**Estado final de la Fase 1:**  Completada y aprobada.

## 4) Feature Engineering
Script: `etl/features/30_generate_features.py`

Se generan, entre otras, `event_count`, `recency_days` y `unique_event_types` a nivel usuario.


In [5]:
feat_path = path + "data/features/user_features.csv"
features = pd.read_csv(feat_path)
print("Feature store:", features.shape)
print("Columnas:", list(features.columns)[:30], "...")
display(features.head(5))


Feature store: (5000, 27)
Columnas: ['user_id', 'signup_date', 'age', 'gender', 'country', 'device', 'os_version', 'app_version', 'language', 'acquisition_channel', 'subscription_type', 'last_active_date', 'sessions_7d', 'views', 'likes', 'avg_watch_time_sec', 'credits_purchased', 'credits_spent', 'episodes_completed', 'top_show_id', 'top_show_title', 'top_show_genre', 'top_show_launch_date', 'top_show_episode_count', 'event_count', 'recency_days', 'unique_event_types'] ...


user_id                signup_date  age  \
0  00009a42-5cfb-413d-ab26-061f5a244411  2025-02-03 00:00:00+00:00   25   
1  001ea567-461e-48bc-976c-b4c35a8cad45  2024-05-13 00:00:00+00:00   31   
2  00242cc9-92ab-406f-a0b2-7f785623c596  2024-05-13 00:00:00+00:00   38   
3  00269d38-ce4d-4ab4-87e3-78bf44683c60  2024-08-07 00:00:00+00:00   42   
4  002cdaea-2b35-4c0a-83c9-710d036a750c  2024-04-15 00:00:00+00:00   24   

  gender        country   device  os_version app_version language  \
0      M       Colombia  Android          14      1.11.3       es   
1      M      Venezuela      iOS          17      1.11.4       es   
2      F       Colombia  Android          12       1.8.2       es   
3      M         Mexico  Android          13      1.14.2       pt   
4      F  United States      iOS          17      1.10.3       es   

  acquisition_channel subscription_type           last_active_date  \
0              tiktok              none  2025-10-05 00:00:00+00:00   
1           instagram              none  2025-07-09 00:00:00+00:00   
2             organic              none  2025-04-15 00:00:00+00:00   
3             organic           premium  2024-10-12 00:00:00+00:00   
4            meta_ads             basic  2025-05-22 00:00:00+00:00   

   sessions_7d  views  likes  avg_watch_time_sec  credits_purchased  \
0            8     11      0                 106                 21   
1            2     16      2                 156                 29   
2            1      7      0                 206                 21   
3            4     12      0                 178                  0   
4            2     18      2                 219                  0   

   credits_spent  episodes_completed                           top_show_id  \
0             19                  12  6a8c74e5-e7ea-40e1-a506-40f9155fba35   
1             29                  16  f4002254-6a02-4e6d-964e-35ce7f8b81c6   
2              3                   0  6a8c74e5-e7ea-40e1-a506-40f9155fba35   
3              0                   8  f4002254-6a02-4e6d-964e-35ce7f8b81c6   
4              0                  18  f4002254-6a02-4e6d-964e-35ce7f8b81c6   

    top_show_title top_show_genre top_show_launch_date  \
0   Código Peligro         Acción           2025-04-02   
1  Destino Cruzado          Drama           2025-02-14   
2   Código Peligro         Acción           2025-04-02   
3  Destino Cruzado          Drama           2025-02-14   
4  Destino Cruzado          Drama           2025-02-14   

   top_show_episode_count  event_count  recency_days  unique_event_types  
0                      38           38            52                   7  
1                      42           36           106                   7  
2                      38           30           198                   7  
3                      42           30            89                   6  
4                      42           36           157                   7

## 5) Cohortes y retención
Script: `etl/cohorts/31_cohorts_retention.py`

Cohortes por `signup_date` mensual y actividad por `event_month`. Se calcula `retention_rate`.


In [6]:
cohorts_path = path + "data/cohorts/retention_cohorts.csv"
retention = pd.read_csv(cohorts_path)
print("Cohortes de retención:", retention.shape)
display(retention.head(10))

n_cohorts = retention['cohort_month'].nunique()
print("Número de cohortes:", n_cohorts)


Cohortes de retención: (184, 5)


cohort_month event_month  active_users  cohort_size  retention_rate
0      2024-01     2025-01           469          734        0.638965
1      2024-01     2025-02           429          734        0.584469
2      2024-01     2025-03           450          734        0.613079
3      2024-01     2025-04           425          734        0.579019
4      2024-01     2025-05           433          734        0.589918
5      2024-01     2025-06           425          734        0.579019
6      2024-01     2025-07           373          734        0.508174
7      2024-01     2025-08           307          734        0.418256
8      2024-01     2025-09           225          734        0.306540
9      2024-01     2025-10            77          734        0.104905

Número de cohortes: 22


## 6) Modelado predictivo
Script: `etl/modeling/41_train_churn_model.py`

Se entrenaron dos modelos:
- LogisticRegression (class_weight='balanced')
- RandomForestClassifier (class_weight='balanced')

El resumen de métricas se persiste en `docs/model_eval_summary.json`.


In [7]:
eval_json = path + "docs/model_eval_summary.json"
if os.path.exists(eval_json):
    with open(eval_json, "r", encoding="utf-8") as f:
        summary = json.load(f)
    print("Resumen de evaluación de modelos:")
    display(pd.DataFrame(summary).T)
else:
    print("No se encontró docs/model_eval_summary.json. Ejecuta 41_train_churn_model.py para generarlo.")


Resumen de evaluación de modelos:


AUC  Recall  Precision        F1  Threshold
logreg         0.963730     1.0      0.848  0.917749   0.070787
random_forest  0.755481     1.0      0.848  0.917749   0.362500

## 7) Scoring de usuarios
Script: `etl/modeling/42_predict_churn.py`

Se generan `proba_churn` y `pred_churn` por usuario en `data/models/predictions.csv`.


In [9]:
preds_path = path + "data/models/predictions_full.csv"
if os.path.exists(preds_path):
    preds = pd.read_csv(preds_path)
    print("Predicciones:", preds.shape)
    display(preds.head(10)[["user_id","proba_churn","pred_churn"]])
    print("Distribución de proba_churn:")
    display(preds["proba_churn"].describe().to_frame("proba_churn"))
else:
    print("No se encontró data/models/predictions.csv. Ejecuta 42_predict_churn.py para generarlo.")


Predicciones: (5000, 6)


user_id  proba_churn  pred_churn
0  00009a42-5cfb-413d-ab26-061f5a244411          1.0           1
1  001ea567-461e-48bc-976c-b4c35a8cad45          1.0           1
2  00242cc9-92ab-406f-a0b2-7f785623c596          1.0           1
3  00269d38-ce4d-4ab4-87e3-78bf44683c60          1.0           1
4  002cdaea-2b35-4c0a-83c9-710d036a750c          1.0           1
5  003500a5-7fb7-4175-99bc-fc03f0ff0cf6          1.0           1
6  004786bb-6e4d-49b1-989a-1db1f067ae0b          1.0           1
7  0050fd08-6f9f-40e0-b562-42d0a2ab8005          1.0           1
8  0053cf9f-8090-48e0-8c1d-a4f6bb01f0a1          1.0           1
9  005bb6b9-1722-4eb6-bffd-9100c5b8107b          1.0           1

Distribución de proba_churn:


proba_churn
count  5000.000000
mean      0.990931
std       0.068652
min       0.015404
25%       0.999997
50%       1.000000
75%       1.000000
max       1.000000

## 8) QA y calibración de scoring
Script: `etl/analysis/43_churn_scoring_QA.py`

Se revisan distribuciones agregadas y segmentación por percentiles de riesgo.


In [12]:
docs_dir = path + "docs/"
by_country = docs_dir + "churn_by_country.csv"
by_device  = docs_dir + "churn_by_device.csv"
by_subs    = docs_dir + "churn_by_subscription.csv"
seg_path   = docs_dir + "churn_segmented.csv"

def maybe_show(p, title):
    if os.path.exists(p):
        df = pd.read_csv(p)
        print(title, "|", os.path.basename(p), "| shape:", df.shape)
        display(df.head(10))
    else:
        print("No existe:", p)

maybe_show(by_country, "Promedio de probabilidad por país")
maybe_show(by_device,  "Promedio de probabilidad por dispositivo")
maybe_show(by_subs,    "Promedio de probabilidad por tipo de suscripción")
maybe_show(seg_path,   "Segmentación por percentiles de riesgo")


Promedio de probabilidad por país | churn_by_country.csv | shape: (8, 2)


country  avg_churn_prob
0           Peru        0.998258
1          Chile        0.995511
2      Venezuela        0.995059
3          Spain        0.992301
4         Mexico        0.990822
5  United States        0.990535
6       Colombia        0.990029
7      Argentina        0.978380

Promedio de probabilidad por dispositivo | churn_by_device.csv | shape: (2, 2)


device  avg_churn_prob
0      iOS        0.992313
1  Android        0.990321

Promedio de probabilidad por tipo de suscripción | churn_by_subscription.csv | shape: (3, 2)


subscription_type  avg_churn_prob
0             basic        0.992472
1              none        0.990886
2           premium        0.987564

Segmentación por percentiles de riesgo | churn_segmented.csv | shape: (5000, 32)


user_id  proba_churn  pred_churn  \
0  00009a42-5cfb-413d-ab26-061f5a244411          1.0           1   
1  001ea567-461e-48bc-976c-b4c35a8cad45          1.0           1   
2  00242cc9-92ab-406f-a0b2-7f785623c596          1.0           1   
3  00269d38-ce4d-4ab4-87e3-78bf44683c60          1.0           1   
4  002cdaea-2b35-4c0a-83c9-710d036a750c          1.0           1   
5  003500a5-7fb7-4175-99bc-fc03f0ff0cf6          1.0           1   
6  004786bb-6e4d-49b1-989a-1db1f067ae0b          1.0           1   
7  0050fd08-6f9f-40e0-b562-42d0a2ab8005          1.0           1   
8  0053cf9f-8090-48e0-8c1d-a4f6bb01f0a1          1.0           1   
9  005bb6b9-1722-4eb6-bffd-9100c5b8107b          1.0           1   

   threshold_used              model_file        scored_at_utc  \
0        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
1        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
2        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
3        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
4        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
5        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
6        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
7        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
8        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   
9        0.070787  churn_model_logreg.pkl  2025-10-27T11:24:08   

                 signup_date  age gender        country   device  os_version  \
0  2025-02-03 00:00:00+00:00   25      M       Colombia  Android          14   
1  2024-05-13 00:00:00+00:00   31      M      Venezuela      iOS          17   
2  2024-05-13 00:00:00+00:00   38      F       Colombia  Android          12   
3  2024-08-07 00:00:00+00:00   42      M         Mexico  Android          13   
4  2024-04-15 00:00:00+00:00   24      F  United States      iOS          17   
5  2024-03-20 00:00:00+00:00   36      M       Colombia  Android          14   
6  2024-01-16 00:00:00+00:00   28      M          Spain  Android          14   
7  2025-04-13 00:00:00+00:00   23      F         Mexico  Android          12   
8  2024-10-10 00:00:00+00:00   18      F          Spain  Android          12   
9  2024-03-10 00:00:00+00:00   34      F       Colombia      iOS          15   

  app_version language acquisition_channel subscription_type  \
0      1.11.3       es              tiktok              none   
1      1.11.4       es           instagram              none   
2       1.8.2       es             organic              none   
3      1.14.2       pt             organic           premium   
4      1.10.3       es            meta_ads             basic   
5      1.13.3       es             organic           premium   
6      1.13.1       es             organic              none   
7      1.13.3       es             organic              none   
8      1.12.2       es              tiktok           premium   
9      1.13.4       en              tiktok              none   

            last_active_date  sessions_7d  views  likes  avg_watch_time_sec  \
0  2025-10-05 00:00:00+00:00            8     11      0                 106   
1  2025-07-09 00:00:00+00:00            2     16      2                 156   
2  2025-04-15 00:00:00+00:00            1      7      0                 206   
3  2024-10-12 00:00:00+00:00            4     12      0                 178   
4  2025-05-22 00:00:00+00:00            2     18      2                 219   
5  2025-08-30 00:00:00+00:00            1     14      1                  27   
6  2025-09-06 00:00:00+00:00            3     15      0                 257   
7  2025-06-25 00:00:00+00:00            1     19      1                 218   
8  2025-10-09 00:00:00+00:00            1      6      1                  28   
9  2025-10-10 00:00:00+00:00            0     15      2                 156   

   credits_purchased  credits_spent  episodes_completed  \
0                 21             19      

## 9) Segmentación no supervisada (K-Means)
Script: `etl/segmentation/51_user_segmentation.py`

Se seleccionan variables de actividad para clustering, se evalúa `k` por Silhouette y se genera `user_clusters.csv`.


In [13]:
clusters_path = path + "data/analytics/user_clusters.csv"
clusters_summary_path = path + "docs/cluster_summary.csv"

if os.path.exists(clusters_path):
    cl = pd.read_csv(clusters_path)
    print("Asignaciones de cluster:", cl.shape)
    display(cl.head(10))
    print("Distribución de clusters:")
    display(cl["cluster"].value_counts(normalize=True).mul(100).round(2).to_frame("%"))
else:
    print("No se encontró data/analytics/user_clusters.csv. Ejecuta 51_user_segmentation.py.")

if os.path.exists(clusters_summary_path):
    cs = pd.read_csv(clusters_summary_path)
    print("Resumen de clusters:", cs.shape)
    display(cs.head(10))
else:
    print("No se encontró docs/cluster_summary.csv. Ejecuta 51_user_segmentation.py.")


Asignaciones de cluster: (5000, 28)


user_id                signup_date  age  \
0  00009a42-5cfb-413d-ab26-061f5a244411  2025-02-03 00:00:00+00:00   25   
1  001ea567-461e-48bc-976c-b4c35a8cad45  2024-05-13 00:00:00+00:00   31   
2  00242cc9-92ab-406f-a0b2-7f785623c596  2024-05-13 00:00:00+00:00   38   
3  00269d38-ce4d-4ab4-87e3-78bf44683c60  2024-08-07 00:00:00+00:00   42   
4  002cdaea-2b35-4c0a-83c9-710d036a750c  2024-04-15 00:00:00+00:00   24   
5  003500a5-7fb7-4175-99bc-fc03f0ff0cf6  2024-03-20 00:00:00+00:00   36   
6  004786bb-6e4d-49b1-989a-1db1f067ae0b  2024-01-16 00:00:00+00:00   28   
7  0050fd08-6f9f-40e0-b562-42d0a2ab8005  2025-04-13 00:00:00+00:00   23   
8  0053cf9f-8090-48e0-8c1d-a4f6bb01f0a1  2024-10-10 00:00:00+00:00   18   
9  005bb6b9-1722-4eb6-bffd-9100c5b8107b  2024-03-10 00:00:00+00:00   34   

  gender        country   device  os_version app_version language  \
0      M       Colombia  Android          14      1.11.3       es   
1      M      Venezuela      iOS          17      1.11.4       es   
2      F       Colombia  Android          12       1.8.2       es   
3      M         Mexico  Android          13      1.14.2       pt   
4      F  United States      iOS          17      1.10.3       es   
5      M       Colombia  Android          14      1.13.3       es   
6      M          Spain  Android          14      1.13.1       es   
7      F         Mexico  Android          12      1.13.3       es   
8      F          Spain  Android          12      1.12.2       es   
9      F       Colombia      iOS          15      1.13.4       en   

  acquisition_channel subscription_type  sessions_7d  views  likes  \
0              tiktok              none            8     11      0   
1           instagram              none            2     16      2   
2             organic              none            1      7      0   
3             organic           premium            4     12      0   
4            meta_ads             basic            2     18      2   
5             organic           premium            1     14      1   
6             organic              none            3     15      0   
7             organic              none            1     19      1   
8              tiktok           premium            1      6      1   
9              tiktok              none            0     15      2   

   avg_watch_time_sec  credits_purchased  credits_spent  episodes_completed  \
0                 106                 21             19                  12   
1                 156                 29             29                  16   
2                 206                 21              3                   0   
3                 178                  0              0                   8   
4                 219                  0              0                  18   
5                  27                  0              0                  11   
6                 257                  8              8                  17   
7                 218                 18             11                   9   
8                  28                 26             26                   3   
9                 156                 49             38                  15   

                            top_show_id       top_show_title top_show_genre  \
0  6a8c74e5-e7ea-40e1-a506-40f9155fba35       Código Peligro         Acción   
1  f4002254-6a02-4e6d-964e-35ce7f8b81c6      Destino Cruzado          Drama   
2  6a8c74e5-e7ea-40e1-a506-40f9155fba35       Código Peligro         Acción   
3  f4002254-6a02-4e6d-964e-35ce7f8b81c6      Destino Cruzado          Drama   
4  f4002254-6a02-4e6d-964e-35ce7f8b81c6      Destino Cruzado          Drama   
5  6c110fa9-2275-49c5-b0da-a635dcc5c65a  Corazones en Llamas        Romance   
6  98eff27c-e737-46ec-9722-8cc78c7b5b0d           Dulce Vida        Romance   
7  98eff27c-e737-46ec-9722-8cc78c7b5b0d           Dulce Vida        Romance   
8  47b511a2-d290-4866-a496-bc76a4d7f721   Sombra de Venganza       Thriller   
9  6c110fa9-227

Distribución de clusters:


%
cluster       
1        61.16
0        38.84

Resumen de clusters: (2, 11)


cluster  event_count  recency_days  unique_event_types  credits_purchased  \
0        0        40.74         64.76                6.92              30.49   
1        1        39.53         80.57                6.92               6.86   

   credits_spent  views  avg_watch_time_sec  sessions_7d  user_count  \
0          21.23  16.32              142.31         3.23        1942   
1           2.51  11.92              101.58         2.09        3058   

   pct_total  
0      38.84  
1      61.16

## 10) Visualización de clusters
Script: `etl/segmentation/52_cluster_visuals.py`

Los gráficos se guardan en `docs/`. Si no aparecen, verificar librerías gráficas del entorno.


![alt text](../docs/cluster_boxplot_views.png)
![alt text](../docs/cluster_pca_scatter.png)

In [16]:
expected_imgs = [
    "cluster_pca_scatter.png",
    "cluster_boxplot_views.png"
]
for img in expected_imgs:
    p = path + "docs/" + img
    print("Existe imagen:", img, "->", os.path.exists(p), "|", p)


Existe imagen: cluster_pca_scatter.png -> True | C:/Users/Setoro/Desktop/Idilio/IdilioTv/docs/cluster_pca_scatter.png
Existe imagen: cluster_boxplot_views.png -> True | C:/Users/Setoro/Desktop/Idilio/IdilioTv/docs/cluster_boxplot_views.png


# Executive Summary

1. El comportamiento agregado muestra una base con alta proporción de usuarios que no regresan en 30 días según la definición operacional (churn_30d), con fuerte concentración de probabilidades de churn cercanas a 1.0 por la combinación de recencia alta y baja diversidad de eventos en la mayoría de usuarios.  
2. La curva ROC de la regresión logística es alta (AUC ≈ 0.96), pero la calibración no es confiable para leer probabilidades como riesgo absoluto. Se usa el modelo como ranker y se recomienda calibración posterior (Platt/Isotónica) con muestreo temporal apropiado.  
3. Cohortes: tamaños similares, retención decay típica M0→M1→M2; variaciones por país y dispositivo justifican pruebas localizadas.  
4. Segmentación K-Means sugiere 2 perfiles: uno más activo con mayor gasto y vistas, y otro más inactivo con peor recencia y menor diversidad. Esto permite acciones diferenciadas.

Recomendaciones:
- Orquestar experimentos con threshold por segmento. Medir lift por decil de probabilidad y comparar contra baseline.  
- Activar campañas por país/dispositivo con disparadores de recencia y play intent fallido.  
- Calibrar el modelo y añadir features de sesión (p95 duración, exit_rate) y señales de monetización (ARPU/ARPPU).  
- Preparar canalizaciones de dashboards con KPIs recurrentes (D7/D30, churn, ratio de plays/sesión, next_rate), y una tabla de riesgo actualizable.
